core/utils/promotions/filtered-promotions-manager.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Elementor\Core\Utils\Promotions;
+
+use function DI\string;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Filtered_Promotions_Manager {
+
+	/**
+	 * @param array $promotion_data
+	 * @param string $filter_name
+	 * @param string $url_key
+	 * @return array
+	 */
+	public static function get_filtered_promotion_data( array $promotion_data, string $filter_name, string $url_key, string $url_sub_key = '' ): array {
+		$new_promotion_data = apply_filters( $filter_name, $promotion_data );
+		if ( ! is_array( $new_promotion_data ) ) {
+			return $promotion_data;
+		}
+
+		$new_promotion_data = array_intersect_key( $new_promotion_data, array_flip( array_keys( $promotion_data ) ) );
```

Comment:
Just a suggestion.
Maybe we can extract this functionality to a new function to improve readability. 
Maybe `clean_promotion_data` or `retain_original_keys`